### ISSUE TITLE
Is there a way to have a batch of jobs?

### ISSUE THREAD

**Issue by nathanl** [koudelka/honeydew#83]
I'd like to do something like this:

- Start a job and include some kind of batch id
- At the end of its work, the job does a check: based on its results, do any other jobs need to be added to the batch?
  - If so, add them
  - If not, see if there are any jobs still queued for this batch
    - If there are still jobs queued for the batch, do nothing
    - If there are no more jobs queued for the batch, fire off a message saying "the batch is done"

Is there any straightforward way to do this? (FWIW I'm using an Mnesia queue.)

**Comment by koudelka** [comment]
Hey there Nathan,

If I'm understanding correctly, the root of your need is to be able to track the status of a dynamically growing/shrinking group of jobs, and when all of them are finished, send a message.

It's certainly possible to check the state of the queue with `Honeydew.filter/1` and filtering for a batch id argument that you'd pass to jobs as part of their task (and subsequent jobs that it enqueues), but I believe that's going to be prone to race conditions.

In order to ensure that you're checking for the condition you truly care about (the job completed vs enqueued/in-progress), I'd suggest an external batch monitoring process.

Since I have poor self control, I've taken the liberty of sketching out such a thing, as well as a little simulator of your scenario. :)

```elixir
defmodule BatchMonitor do
  @moduledoc """
  A quick and dirty job batch monitor.

  When you enqueue a job, call `track_id/2` to begin tracking the job id,
  you'll then need to include that id in the job's task specification.

  Honeydew's Success/Failure Modes are used to determine when the job completes.
  """
  use GenServer

  require Logger

  alias Honeydew.Job

  #
  # Success/Failure Mode callbacks
  #

  defmodule SuccessMode do
    @behaviour Honeydew.SuccessMode

    @impl true
    def validate_args!([]), do: :ok

    @impl true
    def handle_success(%Job{task: {_function, [%{id: id, batch_monitor: batch_monitor}]}}, []) do
      BatchMonitor.report_success(batch_monitor, id)
    end
  end

  defmodule FailureMode do
    @behaviour Honeydew.FailureMode

    @impl true
    def validate_args!([]), do: :ok

    @impl true
    def handle_failure(%Job{task: {_function, [%{id: id, batch_monitor: batch_monitor}]}}, _reason, []) do
      BatchMonitor.report_failure(batch_monitor, id)
    end
  end


  #
  # Client API
  #

  def start_link(batch_id) do
    GenServer.start_link(__MODULE__, batch_id)
  end

  def track_id(batch_monitor, id) do
    GenServer.call(batch_monitor, {:track, id})
  end

  def report_success(batch_monitor, id) do
    GenServer.call(batch_monitor, {:succeeded, id})
  end

  def report_failure(batch_monitor, id) do
    GenServer.call(batch_monitor, {:failed, id})
  end


  #
  # GenServer callbacks
  #

  def init(batch_id) do
    log batch_id, "Started batch monitor."

    {:ok, {batch_id, MapSet.new()}}
  end

  def handle_call({:track, id}, _from, {batch_id, ids}) do
    log batch_id, "Tracking Job #{id}."

    {:reply, :ok, {batch_id, MapSet.put(ids, id)}}
  end

  def handle_call({:succeeded, id}, _from, {batch_id, _ids} = state) do
    log batch_id, "Job #{id} succeeded."

    remove_id_and_maybe_stop(state, id)
  end

  def handle_call({:failed, id}, _from, {batch_id, _ids} = state) do
    log batch_id, "Job #{id} failed."

    remove_id_and_maybe_stop(state, id)
  end

  defp remove_id_and_maybe_stop({batch_id, ids}, id) do
    ids = remove_id(ids, id)

    if batch_complete?(ids) do
      log batch_id, "Batch complete."
      {:stop, :normal, :ok, {batch_id, ids}}
    else
      {:reply, :ok, {batch_id, ids}}
    end
  end

  defp remove_id(ids, id) do
    MapSet.delete(ids, id)
  end

  defp batch_complete?(ids) do
    MapSet.size(ids) == 0
  end

  defp log(batch_id, msg) do
    Logger.info "[Batch #{batch_id}] #{msg}"
  end
end


defmodule Worker do
  @behaviour Honeydew.Worker

  require Logger

  def run(%{job_arg: job_arg, batch_monitor: batch_monitor}) do
    need_to_add_jobs = :rand.uniform() > 0.5

    if need_to_add_jobs do
      num_to_add = 1

      Enum.each(1..num_to_add + 1, &Worker.enqueue(job_arg + &1, batch_monitor))
    end

    should_fail = :rand.uniform() > 0.8

    if should_fail do
      raise "intentional failure!"
    end
  end

  def enqueue(job_arg, batch_monitor) do
    id = :erlang.unique_integer([:monotonic])
    BatchMonitor.track_id(batch_monitor, id)

    {:run, [%{id: id, job_arg: job_arg, batch_monitor: batch_monitor}]}
    |> Honeydew.async(queue())
  end

  def queue do
    :my_queue
  end
end


defmodule App do
  def start do
    queue = Worker.queue()
    :ok = Honeydew.start_queue(queue, success_mode: BatchMonitor.SuccessMode,
                                      failure_mode: BatchMonitor.FailureMode)
    :ok = Honeydew.start_workers(queue, Worker, num: 10)
  end
end


App.start

job_arg = 1
batch_id = "1234"

{:ok, batch_monitor} = BatchMonitor.start_link(batch_id)

Worker.enqueue(job_arg, batch_monitor)
```

Among other things, you'll probably want to give some thought to supervising the batch monitor and what to do if it dies (I'd suggest restarting the "root" job).

In the above, I'm passing around the `batch_monitor` pid, you could avoid that by using a process registry with the batch id, but there's going to be a tradeoff there if you've got workers on multiple machines.

Unfortunately there's no transactional integrity between tracking an id and the job's presence in the queue, for that, the batch monitor would need to live inside the queue implementation itself. That could be something worth including in the library in the future.

**Comment by nathanl** [comment]
Awesome, thanks for the input. The project currently uses spawns within the job and awaits them. After our next milestone I'll probably take a shot at implementing something like this.

**Comment by nathanl** [comment]
I'm tinkering with this. I would want my `SuccessMode` to do different things based on the return value of the function.

From https://github.com/koudelka/honeydew/blob/master/lib/honeydew/job_runner.ex#L71 it appears that the `%Honeydew.Job` handed to `SuccessMode.handle_success` would include the function's return value as `result:`, but it comes back `nil` for me, even though logging at that point shows the right value.

**Comment by koudelka** [comment]
Hey sorry about that, I've just published version `1.4.4` to fix it.

The success mode callback is executed by the "job monitor" process, in order to free the worker up immediately, but the result wasn't getting passed back to the monitor. Should be all set now. :)

**Comment by nathanl** [comment]
Yep, that got the `result` back correctly! Thanks! 😄
